src/pybind/rbd/rbd.pyx

```diff
@@ -444,6 +465,122 @@ cdef void* realloc_chk(void* ptr, size_t size) except NULL:
         raise MemoryError("realloc failed")
     return ret
 
+cdef class Completion
+
+cdef void __aio_complete_cb(rbd_completion_t completion, void *args) with gil:
+    """
+    Callback to oncomplete() for asynchronous operations
+    """
+    cdef Completion cb = <Completion>args
+    cb._complete()
+
+
+@cython.no_gc_clear
+cdef class Completion(object):
+    """completion object"""
+
+    cdef public:
+         object image
+         object oncomplete
+         object ref
+         object event
+         object exc_info
+
+    cdef:
+         rbd_completion_t rbd_comp
+         PyObject* buf
+
+    def __cinit__(self, image, object oncomplete):
+        self.oncomplete = oncomplete
+        self.image = image
+        self.event = threading.Event()
+
+    def is_complete(self):
+        """
+        Has an asynchronous operation completed?
+
+        This does not imply that the callback has finished.
+
+        :returns: True if the operation is completed
+        """
+        with nogil:
+            ret = rbd_aio_is_complete(self.rbd_comp)
+        return ret == 1
+
+    def wait_for_complete(self):
+        """
+        Wait for an asynchronous operation to complete
+
+        This does not imply that the complete callback has finished.
+        """
+        with nogil:
+            rbd_aio_wait_for_complete(self.rbd_comp)
+
+    def wait_for_complete_and_cb(self):
```

Comment:
I'm not really familiar with the nuances of the async handling in librbd, but if you can't think of a reason for the `done = true` to happen before the callback I can send a PR to move that back and get rid of the event. Let me know if that sounds sensible.